server.go

```diff
@@ -412,6 +413,36 @@ func (c *serverConn) run(sctx context.Context) {
 		}
 	}(recvErr)
 
+	go func(responseErr chan error) {
+		for {
+			select {
+			// We don't want a case for c.shutdown here, as that would cause us to exit
+			// immediately when it is signaled, rather than waiting for any active requests
+			// to complete first. Instead, once all the active requests have completed,
+			// the main loop will return and close done, which will cause us to exit as well.
+			case <-done:
+				return
+			case response := <-responses:
```

Comment:
May be slightly clearer to `defer close(responses)` and just have this be `for response := range responses`.

Reply:
Ah, maybe that's not practical since responses might still be referenced in the `call` goroutine.

Reply:
Yes, generally you don't want to close from the read side.